datastream-server-restli/src/main/java/com/linkedin/datastream/server/dms/DatastreamSourceCheckpointResources.java

```diff
@@ -0,0 +1,146 @@
+
+
+    /**
+     * Construct an instance of DatastreamCheckpointResources
+     * @param datastreamServer Datastream server for which health data is retrieved
+     */
+    public DatastreamSourceCheckpointResources(DatastreamServer datastreamServer) {
+        _server = datastreamServer;
+        _store = _server.getDatastreamStore();
+        _coordinator = _server.getCoordinator();
+        _errorLogger = new ErrorLogger(LOG, _coordinator.getInstanceName());
+    }
+
+    /**
+     * Get source checkpoints of the parent datastream for each tasks
+     * @return current checkpoint
+     */
+    @Override
+    public ConnectorHealth get() {
+        String datastreamName = getParentKey();
+        try {
+            Datastream datastream = _store.getDatastream(datastreamName);
+            if (datastream == null) {
+                _errorLogger.logAndThrowRestLiServiceException(HttpStatus.S_404_NOT_FOUND,
+                        "Datastream does not exist: " + datastreamName);
+
+            }
+            return buildSourceCheckpoint(datastream);
+        } catch (Exception e) {
+            _errorLogger.logAndThrowRestLiServiceException(HttpStatus.S_500_INTERNAL_SERVER_ERROR,
+                    "Get datastream checkpoint failed for datastream: " + datastreamName, e);
+        }
+        // Returning null will automatically trigger a 404 Not Found response
+        return null;
+    }
+
+    /**
+     * Update source checkpoint on a given datastream
+     * @param input desired checkpoint value
+     * @return
+     */
+    @Override
+    public UpdateResponse update(ConnectorHealth input) {
+        String datastreamName = getParentKey();
+        Datastream datastream = _store.getDatastream(datastreamName);
+        if (datastream == null) {
+            _errorLogger.logAndThrowRestLiServiceException(HttpStatus.S_404_NOT_FOUND,
+                    "Datastream does not exist: " + datastreamName);
+        }
+        // Currently only KafkaCustomCheckpointProvider supports rewinding the checkpoint
+        if (!_server.isCustomCheckpointing(datastream.getConnectorName())) {
+            _errorLogger.logAndThrowRestLiServiceException(HttpStatus.S_501_NOT_IMPLEMENTED,
+                    "This datastream doesn't support checkpoints to be updated.");
+        }
+        // Only paused datastreams can be updated
+        if (!DatastreamStatus.PAUSED.equals(datastream.getStatus())) {
+            _errorLogger.logAndThrowRestLiServiceException(HttpStatus.S_405_METHOD_NOT_ALLOWED,
+                    "Datastream must be in PAUSED state before updating checkpoint");
+        }
+        try {
+            // Sample SourceCheckpoint resource {"tasks":["sourceCheckpoint":"1234"]}
+            final long newCheckpoint = Long.parseLong(input.getTasks().get(0).getSourceCheckpoint());
+            KafkaCustomCheckpointProvider kafkaCustomCheckpointProvider = (KafkaCustomCheckpointProvider) _server.getCustomCheckpointProvider(datastream);
```

Comment:
You don't need to cast it to KafkaCustomCheckpointProvider. CustomCheckpointProvider defines the methods you need.